From cd1f0e39420c1fbee54f7a22e8bd57a48057951e Mon Sep 17 00:00:00 2001
From: Grzegorz Jaszczyk <jaz@semihalf.com>
Date: Thu, 12 Jan 2017 15:50:55 +0100
Subject: [PATCH 21/45] usb: gadget: u3d: allow to request both threaded and
 and not-threaded irq

The problem with previous request is that it works fine with direct gpio's
but fails when gpio from io-expander (pca953x) are registered. The gpio
from io-expander should be requested with the 'threaded' version of
request_irq, since they are nested interrupts. Using function
devm_request_any_context_irq for requesting gpio interrupts solve described
issue, since it selects either hardirq or threaded handling method
depending on the context.

Change-Id: I5a56d29e606b633cdbbdcb6c4af2b1a6cdd597c3
Signed-off-by: Grzegorz Jaszczyk <jaz@semihalf.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/38532
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 drivers/usb/gadget/udc/mvebu_u3d.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/gadget/udc/mvebu_u3d.c b/drivers/usb/gadget/udc/mvebu_u3d.c
index 3313f13e030c..147ab5e19a58 100644
--- a/drivers/usb/gadget/udc/mvebu_u3d.c
+++ b/drivers/usb/gadget/udc/mvebu_u3d.c
@@ -2348,12 +2348,16 @@ static int mvc2_probe(struct platform_device *pdev)
 	if (gpio_is_valid(cp->vbus_pin)) {
 		cp->prev_vbus = 0;
 		if (!devm_gpio_request(&pdev->dev, cp->vbus_pin, "mvebu-u3d")) {
-			ret = devm_request_irq(&pdev->dev,
+			/* Use the 'any_context' version of function to allow
+			 * requesting both direct GPIO interrupt (hardirq) and
+			 * IO-expander's GPIO (nested interrupt)
+			 */
+			ret = devm_request_any_context_irq(&pdev->dev,
 					       gpio_to_irq(cp->vbus_pin),
 					       mvc2_vbus_irq,
 					       IRQ_TYPE_EDGE_BOTH, "mvebu-u3d",
 					       cp);
-			if (ret) {
+			if (ret < 0) {
 				cp->vbus_pin = -ENODEV;
 				dev_warn(&pdev->dev,
 					 "failed to request vbus irq; "
-- 
2.35.3

